test/test_charm.py

```diff
@@ -697,262 +685,273 @@ def test_containers_storage_multiple_mounts(self):
+
+    framework = create_framework(request)
+    charm = MyCharm(framework)
+
+    charm.on.secret_changed.emit('secret:changed', None)
+    charm.on.secret_rotate.emit('secret:rotate', 'rot')
+    charm.on.secret_remove.emit('secret:remove', 'rem', 7)
+    charm.on.secret_expired.emit('secret:expired', 'exp', 42)
+
+    assert charm.seen == [
+        'SecretChangedEvent',
+        'SecretRotateEvent',
+        'SecretRemoveEvent',
+        'SecretExpiredEvent',
+    ]
+
+
+def test_collect_app_status_leader(request: pytest.FixtureRequest, fake_script: FakeScript):
+    class MyCharm(ops.CharmBase):
+        def __init__(self, *args: typing.Any):
+            super().__init__(*args)
+            self.framework.observe(self.on.collect_app_status, self._on_collect_status)
+
+        def _on_collect_status(self, event: ops.CollectStatusEvent):
+            event.add_status(ops.ActiveStatus())
+            event.add_status(ops.BlockedStatus('first'))
+            event.add_status(ops.WaitingStatus('waiting'))
+            event.add_status(ops.BlockedStatus('second'))
+
+    fake_script.write('is-leader', 'echo true')
+    fake_script.write('status-set', 'exit 0')
+
+    framework = create_framework(request)
+    charm = MyCharm(framework)
+    ops.charm._evaluate_status(charm)
+
+    assert fake_script.calls(True) == [
+        ['is-leader', '--format=json'],
+        ['status-set', '--application=True', 'blocked', 'first'],
+    ]
+
+
+def test_collect_app_status_no_statuses(request: pytest.FixtureRequest, fake_script: FakeScript):
+    class MyCharm(ops.CharmBase):
+        def __init__(self, *args: typing.Any):
+            super().__init__(*args)
+            self.framework.observe(self.on.collect_app_status, self._on_collect_status)
+
+        def _on_collect_status(self, event: ops.CollectStatusEvent):
+            pass
 
-        assert fake_script_calls(self, True) == [
-            ['is-leader', '--format=json'],
-        ]
+    fake_script.write('is-leader', 'echo true')
 
-    def test_collect_app_status_non_leader(self):
-        class MyCharm(ops.CharmBase):
-            def __init__(self, *args: typing.Any):
-                super().__init__(*args)
-                self.framework.observe(self.on.collect_app_status, self._on_collect_status)
+    framework = create_framework(request)
+    charm = MyCharm(framework)
+    ops.charm._evaluate_status(charm)
 
-            def _on_collect_status(self, event: ops.CollectStatusEvent):
-                raise Exception  # shouldn't be called
+    assert fake_script.calls(True) == [
+        ['is-leader', '--format=json'],
+    ]
 
-        fake_script(self, 'is-leader', 'echo false')
 
-        charm = MyCharm(self.create_framework())
-        ops.charm._evaluate_status(charm)
+def test_collect_app_status_non_leader(request: pytest.FixtureRequest, fake_script: FakeScript):
+    class MyCharm(ops.CharmBase):
+        def __init__(self, *args: typing.Any):
+            super().__init__(*args)
```

Comment:
```suggestion
        def __init__(self, framework: ops.Framework):
            super().__init__(framework)
```

There are more of these than I anticipated. Feel free to push back and say we should do it some other time.